tools/gfx/debug-layer.cpp

```diff
@@ -0,0 +1,948 @@
+    }
+}
+
+void DebugCommandBuffer::checkCommandBufferOpenWhenCreatingEncoder()
+{
+    if (!isOpen)
+    {
+        GFX_DIAGNOSE_ERROR("The command buffer is already closed. Encoders can only be retrieved "
+                           "while the command buffer is open.");
+    }
+}
+
+void DebugComputeCommandEncoder::endEncoding()
+{
+    SLANG_GFX_API_FUNC;
+    baseObject->endEncoding();
+}
+
+Result DebugComputeCommandEncoder::bindPipeline(
+    IPipelineState* state,
+    IShaderObject** outRootShaderObject)
+{
+    SLANG_GFX_API_FUNC;
+
+    auto innerState = static_cast<DebugPipelineState*>(state)->baseObject;
+    auto result =
+        baseObject->bindPipeline(innerState, commandBuffer->rootObject.baseObject.writeRef());
+    *outRootShaderObject = &commandBuffer->rootObject;
+    return result;
+}
+
+void DebugComputeCommandEncoder::dispatchCompute(int x, int y, int z)
+{
+    SLANG_GFX_API_FUNC;
+    baseObject->dispatchCompute(x, y, z);
+}
+
+void DebugRenderCommandEncoder::endEncoding()
+{
+    SLANG_GFX_API_FUNC;
+    baseObject->endEncoding();
+}
+
+Result DebugRenderCommandEncoder::bindPipeline(
+    IPipelineState* state,
+    IShaderObject** outRootShaderObject)
+{
+    SLANG_GFX_API_FUNC;
+
+    auto innerState = static_cast<DebugPipelineState*>(state)->baseObject;
+    auto result =
+        baseObject->bindPipeline(innerState, commandBuffer->rootObject.baseObject.writeRef());
+    *outRootShaderObject = &commandBuffer->rootObject;
+    return result;
+}
+
+void DebugRenderCommandEncoder::setViewports(uint32_t count, const Viewport* viewports)
+{
+    SLANG_GFX_API_FUNC;
+    baseObject->setViewports(count, viewports);
+}
+
+void DebugRenderCommandEncoder::setScissorRects(uint32_t count, const ScissorRect* scissors)
+{
+    SLANG_GFX_API_FUNC;
+    baseObject->setScissorRects(count, scissors);
+}
+
+void DebugRenderCommandEncoder::setPrimitiveTopology(PrimitiveTopology topology)
+{
+    SLANG_GFX_API_FUNC;
+    baseObject->setPrimitiveTopology(topology);
+}
+
+void DebugRenderCommandEncoder::setVertexBuffers(
+    UInt startSlot,
+    UInt slotCount,
+    IBufferResource* const* buffers,
+    const UInt* strides,
+    const UInt* offsets)
+{
+    SLANG_GFX_API_FUNC;
+
+    List<IBufferResource*> innerBuffers;
+    for (UInt i = 0; i < slotCount; i++)
+    {
+        innerBuffers.add(static_cast<DebugBufferResource*>(buffers[i])->baseObject.get());
+    }
+    baseObject->setVertexBuffers(startSlot, slotCount, innerBuffers.getBuffer(), strides, offsets);
+}
+
+void DebugRenderCommandEncoder::setIndexBuffer(
+    IBufferResource* buffer,
+    Format indexFormat,
+    UInt offset)
+{
+    SLANG_GFX_API_FUNC;
+    auto innerBuffer = static_cast<DebugBufferResource*>(buffer)->baseObject.get();
+    baseObject->setIndexBuffer(innerBuffer, indexFormat, offset);
+}
+
+void DebugRenderCommandEncoder::draw(UInt vertexCount, UInt startVertex)
+{
+    SLANG_GFX_API_FUNC;
```

Comment:
Obvious opportunity: we could validate here that `draw()` operaitons are only being encoded when a valid graphics pipeline state is in place.